cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusEventPublisher.java

```diff
@@ -208,7 +209,14 @@ private void populateErrorDetailsAndMetrics(final ProgramStatusEventDetails.Buil
       return;
     }
     if (status.isEndState()) {
-      CompletableFuture.supplyAsync(() -> metricsProvider.retrieveMetrics(runId))
+      CompletableFuture.supplyAsync(() -> {
+        try {
+          return metricsProvider.retrieveMetrics(runId);
+        } catch (Exception e) {
+          LOG.error("Error retrieving metrics from provider. ", e);
+          return new ExecutionMetrics[]{};
```

Comment:
I think this is something original JIRA tries to avoid. We should not be calling writeInEventWriters with an empty array.

Reply:
Can you explained more about what's writeInEventWriters means? 
I use return new ExecutionMetrics[]{} here because we still want to get an empty metrics when the fetch fails in retrieveMetrics. (Same as previous logic).

Reply:
Well, reading https://cdap.atlassian.net/browse/CDAP-19283 I was under assumption the whole goal is to skip writing an empty set of metrics if there is an error. Otherwise I don't see what is the problem being solved. @greeshmaswaminathan could you clarify?

Reply:
@tivv , the intent was to not to suppress the exception in the provider and allow the callers to handle it appropriately. In this case, the ProgramStatusEventPublisher would prefer publishing the event even without the metric.